neo/core/spiketrainlist.py

```diff
@@ -0,0 +1,294 @@
+        self._spiketrain_metadata = None
+        self.segment = segment
+
+    def __iter__(self):
+        """Implement iter(self)"""
+        if self._items is None:
+            self._spiketrains_from_array()
+        for item in self._items:
+            yield item
+
+    def __getitem__(self, i):
+        """x.__getitem__(y) <==> x[y]"""
+        if self._items is None:
+            self._spiketrains_from_array()
+        items = self._items[i]
+        if is_spiketrain_or_proxy(items):
+            return items
+        else:
+            return SpikeTrainList(items=items)
+
+    def __str__(self):
+        """Return str(self)"""
+        if self._items is None:
+            if self._spike_time_array is None:
+                return str([])
+            else:
+                return "SpikeTrainList containing {} spikes from {} neurons".format(
+                    self._spike_time_array.size,
+                    len(self._all_channel_ids))
+        else:
+            return str(self._items)
+
+    def __len__(self):
+        """Return len(self)"""
+        if self._items is None:
+            if self._all_channel_ids is not None:
+                return len(self._all_channel_ids)
+            elif self._channel_id_array is not None:
+                return np.unique(self._channel_id_array).size
+            else:
+                return 0
+        else:
+            return len(self._items)
+
+    def _add_spiketrainlists(self, other, in_place=False):
+        if self._spike_time_array is None or other._spike_time_array is None:
+            # if either self or other is not storing multiplexed spike trains
+            # we return
+            if self._items is None:
+                self._spiketrains_from_array()
+            if other._items is None:
+                other._spiketrains_from_array()
+            if in_place:
+                self._items.extend(other._items)
+                return self
+            else:
+                return self.__class__(items=self._items[:] + other._items)
+        else:
+            # both self and other are storing multiplexed spike trains
+            if self._spiketrain_metadata['t_start'] != other._spiketrain_metadata['t_start']:
+                raise ValueError("Incompatible t_start")
+                # todo: adjust times and t_start of other to be compatible with self
+            if self._spiketrain_metadata['t_stop'] != other._spiketrain_metadata['t_stop']:
+                raise ValueError("Incompatible t_stop")
+                # todo: adjust t_stop of self and other as necessary
+            combined_spike_time_array = np.hstack(
+                (self._spike_time_array, other._spike_time_array))
+            combined_channel_id_array = np.hstack(
+                (self._channel_id_array, other._channel_id_array))
+            combined_channel_ids = set(list(self._all_channel_ids) + other._all_channel_ids)
+            if (len(combined_channel_ids) !=
```

Comment:
maybe we should add a `remap_ids` option to automatocally remap ids from the second spiketrainlist

Reply:
that's a convenience behavior to be added later?

Reply:
Clarification: In case of non-empty channel_id overlap the user should have the option to provide a remapping dictionary.

Reply:
This will be added later.